### PR TITLE
Create separate contract and service types with the runtime instance

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

Passing the `ContractRuntime` and the `ServiceRuntime` around the application's methods as arguments is quite verbose.

## Proposal

Create new contract and service types in the applications, and store the runtime there. The original state type is now passed to a `new` required constructor for the contract and service types, and obtained again when it needs to be persisted with a `state_mut` getter.

## Test Plan

Internal refactor, so existing tests should catch regressions.

## Release Plan

Changes the SDK API, so:
- Need to bump the major/minor version number in the next release of the crates.
- Need to update the developer manual.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: examples/fungible/src/contract.rs [hunk 1]
```diff
@@ -60,8 +75,12 @@ impl Contract for FungibleToken {
                 amount,
                 target_account,
             } => {
-                Self::check_account_authentication(None, runtime.authenticated_signer(), owner)?;
-                self.debit(owner, amount).await?;
+                Self::check_account_authentication(
```
Comment (ma2bd) [inline examples/fungible/src/contract.rs:78]: Maybe a non-static method `self.check_account_authentication(..)` is better now

File: examples/native-fungible/src/contract.rs [hunk 2]
```diff
@@ -77,7 +92,7 @@ impl Contract for NativeFungibleToken {
                 target_account,
             } => {
                 Self::check_account_authentication(
```
Comment (ma2bd) [inline examples/native-fungible/src/contract.rs:94]: idem (and probably in other apps?)

**Review by ma2bd — APPROVED** [review]
Awesome!

File: examples/fungible/src/contract.rs (hunk 1 above)
Comment (jvff) [reply examples/fungible/src/contract.rs]: Done, good idea!

File: examples/native-fungible/src/contract.rs (hunk 2 above)
Comment (jvff) [reply examples/native-fungible/src/contract.rs]: Done.

**Comment by jvff** [comment]
As discussed on call this morning, I updated the PR to perform two changes:

The first was to remove the `derive(Default)` from the `ContractRuntime` and the `ServiceRuntime`, adding `pub(crate) fn new() -> Self` constructors instead. Applications now receive the runtime instance (with the proper generic type parameter in the next PR) in their `new()` functions.

The second was to tweak the `linera project new` tests to be not fail in case of warnings from the generated stubbed code.

These newer changes can be seen [here](https://github.com/jvff/linera-protocol/compare/create-separate-contract-and-service-types-r7...create-separate-contract-and-service-types-r7-fixes).

**Review by ma2bd — APPROVED** [review]

File: examples/amm/src/contract.rs [hunk 3]
```diff
@@ -10,28 +10,41 @@ use amm::{AmmError, ApplicationCall, Message, Operation};
...
     contract::system_api,
     ensure, ApplicationCallOutcome, Contract, ContractRuntime, ExecutionOutcome, OutgoingMessage,
     Resources, ViewStateStorage,
 };
 use num_bigint::BigUint;
 use num_traits::{cast::FromPrimitive, ToPrimitive};
 
-linera_sdk::contract!(Amm);
+pub struct AmmContract {
+    state: Amm,
```
Comment (ma2bd) [inline examples/amm/src/contract.rs:22]: Should this be `AmmState`?

File: examples/amm/src/contract.rs [hunk 4]
```diff
@@ -10,28 +10,41 @@ use amm::{AmmError, ApplicationCall, Message, Operation};
...
+impl WithContractAbi for AmmContract {
     type Abi = amm::AmmAbi;
 }
 
 #[async_trait]
-impl Contract for Amm {
+impl Contract for AmmContract {
     type Error = AmmError;
     type Storage = ViewStateStorage<Self>;
+    type State = Amm;
```
Comment (ma2bd) [inline examples/amm/src/contract.rs:36]: Perhaps we could discuss in a separate PR whether the SDK should manage the persistent state at all.

File: examples/amm/src/contract.rs (hunk 3 above)
Comment (jvff) [reply examples/amm/src/contract.rs:22]: Yeah, maybe it makes sense to rename all the state types to have the suffix 🤔

**Pull request merged**